test/optim/test_optim.py

```diff
@@ -60,67 +60,93 @@ def _test_rosenbrock_sparse(
         scheduler_constructors=None,
         sparse_only=False,
         maximize=False,
+        multi_tensor=False
     ):
         if scheduler_constructors is None:
             scheduler_constructors = []
         # For rosenbrock tests, it is mandated that the param is a tensor with 2 numbers
-        param_t = torch.tensor([1.5, 1.5])
+        if multi_tensor:
+            params_t = [torch.tensor([1.5, 1.5]), torch.tensor([1.5, 1.5], dtype=torch.float64)]
+        else:
+            params_t = [torch.tensor([1.5, 1.5])]
 
-        param = Parameter(param_t)
-        optimizer = constructor([param])
+        params = [Parameter(param_t) for param_t in params_t]
+        optimizer = constructor(params)
         schedulers = []
         for scheduler_constructor in scheduler_constructors:
             schedulers.append(scheduler_constructor(optimizer))
 
         if not sparse_only:
-            param_c = Parameter(param_t.clone())
-            optimizer_c = constructor([param_c])
+            params_c = [Parameter(param_t.clone()) for param_t in params_t]
+            optimizer_c = constructor(params_c)
 
         solution = torch.tensor([1, 1])
         with torch.no_grad():
-            initial_dist = param.dist(solution)
+            initial_dist = torch.sum(torch.stack([param.dist(solution) for param in params]))
 
-        def eval(param, sparse_grad, w):
-            # Depending on w, provide only the x or y gradient
-            optimizer.zero_grad()
-            loss = rosenbrock(param)
-            loss.backward()
+        def get_grad(param, sparse_grad):
             grad = drosenbrock(param)
             # NB: We torture test the optimizer by returning an
             # uncoalesced sparse tensor
-            if w:
-                i = torch.LongTensor([[0, 0]])
-                x = grad[0]
-                v = torch.tensor([x / 4.0, x - x / 4.0])
+
+            if sparse_grad:
+                if w:
+                    i = torch.LongTensor([[0, 0]])
+                    x = grad[0]
+                    v = torch.tensor([x / 4.0, x - x / 4.0])
+                else:
+                    i = torch.LongTensor([[1, 1]])
+                    y = grad[1]
+                    v = torch.tensor([y - y / 4.0, y / 4.0])
+                grad_out = torch.sparse_coo_tensor(i, v, (2,), dtype=v.dtype)
             else:
-                i = torch.LongTensor([[1, 1]])
-                y = grad[1]
-                v = torch.tensor([y - y / 4.0, y / 4.0])
-            x = torch.sparse_coo_tensor(i, v, (2,), dtype=v.dtype)
-            with torch.no_grad():
-                if sparse_grad:
-                    param.grad = x
+                if w:
+                    grad_out = torch.tensor([grad[0], 0], dtype=param.dtype)
                 else:
-                    param.grad = x.to_dense()
```

Comment:
Note to reviewer: 

Removing `to_dense` speeds up the test time from 200s to 8s for adagrad. sgd_sparse is still very slow due to optimizer being slow.